### PR TITLE
fix: close #114; use position rather than offset as nextOffset for indexes; close metastream when log is closing

### LINKED ISSUE #114: [E2E] message did not make it to the consumer
```
================================================================================
test_id:    kafkatest.tests.core.reassign_partitions_test.ReassignPartitionsTest.test_reassign_partitions.bounce_brokers=False.reassign_from_offset_zero=False.metadata_quorum=REMOTE_KRAFT
status:     FAIL
run time:   8 minutes 17.514 seconds


    AssertionError("2350 acked message did not make it to the Consumer. They are: 12152, 12153, 12154, 12155, 12156, 12157, 12158, 12159, 12160, 12161, 12162, 12163, 12164, 12165, 12166, 12167, 12168, 12169, 12170, 12171...plus 2330 more. Total Acked: 14666, Total Consumed: 12316. The first 1000 missing messages were validated to ensure they are in Kafka's data files. 1000 were missing. This suggests data loss. Here are some of the messages not found in the data files: [12288, 12289, 12290, 12291, 12292, 12293, 12294, 12295, 12296, 12297]\n")
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/dist-packages/ducktape/tests/runner_client.py", line 186, in _do_run
    data = self.run_test()
  File "/usr/local/lib/python3.9/dist-packages/ducktape/tests/runner_client.py", line 246, in run_test
    return self.test_context.function(self.test)
  File "/usr/local/lib/python3.9/dist-packages/ducktape/mark/_mark.py", line 433, in wrapper
    return functools.partial(f, *args, **kwargs)(*w_args, **w_kwargs)
  File "/opt/kafka-dev/tests/kafkatest/tests/core/reassign_partitions_test.py", line 165, in test_reassign_partitions
    self.run_produce_consume_validate(core_test_action=lambda: self.reassign_partitions(bounce_brokers))
  File "/opt/kafka-dev/tests/kafkatest/tests/produce_consume_validate.py", line 108, in run_produce_consume_validate
    self.validate()
  File "/opt/kafka-dev/tests/kafkatest/tests/produce_consume_validate.py", line 132, in validate
    assert succeeded, error_msg
AssertionError: 2350 acked message did not make it to the Consumer. They are: 12152, 12153, 12154, 12155, 12156, 12157, 12158, 12159, 12160, 12161, 12162, 12163, 12164, 12165, 12166, 12167, 12168, 12169, 12170, 12171...plus 2330 more. Total Acked: 14666, Total Consumed: 12316. The first 1000 missing messages were validated to ensure they are in Kafka's data files. 1000 were missing. This suggests data loss. Here are some of the messages not found in the data files: [12288, 12289, 12290, 12291, 12292, 12293, 12294, 12295, 12296, 12297]


--------------------------------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by mooc9988**

### CODE REVIEW

File: core/src/main/scala/kafka/cluster/Partition.scala [hunk 1]
```diff
@@ -591,6 +591,7 @@ class Partition(val topicPartition: TopicPartition,
       leaderEpochStartOffsetOpt = None
       Partition.removeMetrics(topicPartition)
     }
+    info(s"partition $topicPartition closed, trigger leader re-election")
```
Comment (superhx) [inline core/src/main/scala/kafka/cluster/Partition.scala:594]: - Log already has the ident prefix.
- Add close cost to the log like 'partition closed cost xxms'

File: core/src/main/scala/kafka/log/es/ElasticCheckoutPointFileWithHandler.scala [hunk 2]
```diff
@@ -39,15 +39,22 @@ class ElasticCheckoutPointFileWithHandler(val rawKafkaMeta: RawKafkaMeta) extend
...
+        val (topicPartition, checkpoint) = elem
+        val elasticLog = ElasticLogManager.getElasticLog(topicPartition)
+        if (elasticLog == null) {
+          warn(s"Cannot find elastic log for $topicPartition and skip persistence. Is this broker shutting down?")
+        } else {
+          f(elasticLog, checkpoint)
+        }
       }
+    } catch {
+      case ex: Exception =>
```
Comment (superhx) [inline core/src/main/scala/kafka/log/es/ElasticCheckoutPointFileWithHandler.scala:53]: - Catch throwable may catch more errors.
- Add some utils to deal with the template code.

File: core/src/main/scala/kafka/cluster/Partition.scala (hunk 1 above)
Comment (mooc9988) [reply core/src/main/scala/kafka/cluster/Partition.scala]: solved in the next commit

File: core/src/main/scala/kafka/log/es/ElasticCheckoutPointFileWithHandler.scala (hunk 2 above)
Comment (mooc9988) [reply core/src/main/scala/kafka/log/es/ElasticCheckoutPointFileWithHandler.scala]: solved in the next commit

**Review by superhx — APPROVED** [review]

**Pull request merged**